signer/signer_trust.go

```diff
@@ -85,3 +95,56 @@ func (trust *NotarySigner) GetKey(keyid string) data.PublicKey {
 	}
 	return data.NewPublicKey(data.KeyAlgorithm(publicKey.KeyInfo.Algorithm.Algorithm), publicKey.PublicKey)
 }
+
+// CheckHealth returns true if trust is healthy - that is if it can connect
+// to the trust server, and both the key management and signer services are
+// healthy
+func (trust *NotarySigner) CheckHealth(timeout int) error {
+	if e := trust.checkServiceHealth(
+		"Key manager", trust.kmClient.CheckHealth, timeout); e != nil {
+		return e
+	}
+	return trust.checkServiceHealth(
+		"Signer", trust.sClient.CheckHealth, timeout)
+}
+
+type rpcHealthCheck func(
+	context.Context, *pb.Void, ...grpc.CallOption) (*pb.HealthStatus, error)
+
+// Generalized function that can check the health of both the signer client
+// and the key management client.
+func (trust *NotarySigner) checkServiceHealth(
+	serviceName string, check rpcHealthCheck, timeout int) error {
+
+	// Do not bother starting goroutine if the connection is broken.
+	if trust.clientConn.State() != grpc.Idle &&
+		trust.clientConn.State() != grpc.Ready {
+		return errors.New("Not currently connected to trust server.")
+	}
+
+	// We still want to time out getting health, because the connection could
+	// have disconnected sometime between when we checked the connection and
+	// when we try to make an RPC call.
+	channel := make(chan error)
+	done := make(chan struct{})
+	go func() {
+		status, err := check(context.Background(), &pb.Void{})
+		if len(status.Status) > 0 {
+			err = fmt.Errorf("%s not healthy", serviceName)
+		}
+		select {
+		case <-done:
+		case channel <- err:
+		}
+	}()
+	var returnErr error
+	select {
+	case err := <-channel:
+		returnErr = err
+	case <-time.After(time.Second * time.Duration(timeout)):
+		returnErr = fmt.Errorf("Timed out connecting to %s after %s", serviceName, timeout)
+	}
+	close(done)
+	close(channel)
```

Comment:
I don't think it's correct to close `channel` here. A write to a closed channel panics, and this could try to close `channel` at the same time that the select statement above tries to write to it.